compiler/msgs.nim

```diff
@@ -478,175 +477,246 @@ proc sourceLine*(conf: ConfigRef; i: TLineInfo): string =
-  writeContext(conf, info)
-  liMessage(conf, info, errInternal, errMsg, doAbort, info2)
+  semReportIllformedAst(conf, node, msg)
 
-template internalError*(conf: ConfigRef; info: TLineInfo, errMsg: string) =
-  internalErrorImpl(conf, info, errMsg, instLoc())
+template localReport*(conf: ConfigRef, info: TLineInfo, report: ReportTypes) =
+  handleReport(conf, wrap(report, instLoc(), info), instLoc(), doNothing)
 
-template internalError*(conf: ConfigRef; errMsg: string) =
-  internalErrorImpl(conf, unknownLineInfo, errMsg, instLoc())
+template internalAssert*(
+    conf: ConfigRef, e: bool, failMsg: string = "") =
 
-template internalAssert*(conf: ConfigRef, e: bool) =
-  # xxx merge with `globalAssert`
   if not e:
-    const info2 = instLoc()
-    let arg = info2.toFileLineCol
-    internalErrorImpl(conf, unknownLineInfo, arg, info2)
+    handleReport(
+      conf,
+      wrap(InternalReport(
+        kind: rintAssert, msg: failMsg), instLoc()),
+      instLoc(),
+      doAbort
+    )
+
+template internalError*(
+    conf: ConfigRef, repKind: InternalReportKind, fail: string): untyped =
+  conf.handleReport(
+    wrap(InternalReport(
+      kind: repKind,
+      msg: fail),
+         instLoc()),
+    instLoc(),
+    doAbort
+  )
+
+template internalError*(
+    conf: ConfigRef, info: TLineInfo,
+    repKind: InternalReportKind, fail: string): untyped =
+  conf.handleReport(wrap(
+    InternalReport(
+      kind: repKind, msg: fail), instLoc(), info),
+                    instLoc(), doAbort)
+
+template internalError*(
+    conf: ConfigRef,
+    info: TLineInfo,
+    fail: string,
+  ): untyped =
+
+  conf.handleReport(wrap(
+    InternalReport(kind: rintUnreachable, msg: fail),
+    instLoc(), info), instLoc(), doAbort)
+
+
+template internalError*(
+    conf: ConfigRef,
+    fail: string
+  ): untyped =
+
+  conf.handleReport(wrap(InternalReport(
+    kind: rintUnreachable, msg: fail), instLoc()), instLoc(), doAbort)
 
-template lintReport*(conf: ConfigRef; info: TLineInfo, beau, got: string, forceHint = false, extraMsg = "") =
-  let m = "'$1' should be: '$2'$3" % [got, beau, extraMsg]
-  let msg = if optStyleError in conf.globalOptions and not forceHint: errGenerated else: hintName
-  liMessage(conf, info, msg, m, doNothing, instLoc())
 
 proc quotedFilename*(conf: ConfigRef; i: TLineInfo): Rope =
   if i.fileIndex.int32 < 0:
     result = makeCString "???"
+
   elif optExcessiveStackTrace in conf.globalOptions:
     result = conf.m.fileInfos[i.fileIndex.int32].quotedFullName
+
   else:
     result = conf.m.fileInfos[i.fileIndex.int32].quotedName
 
-template listMsg(title, r) =
-  msgWriteln(conf, title, {msgNoUnitSep})
-  for a in r: msgWriteln(conf, "  [$1] $2" % [if a in conf.notes: "x" else: " ", $a], {msgNoUnitSep})
+proc listWarnings*(conf: ConfigRef) =
+  conf.localReport(InternalReport(
+    kind: rintListWarnings,
+    enabledOptions: repWarningKinds * conf.notes))
```

Comment:
I don't think it's here, but somewhere along the way the `computeNotesVerbosity` related information that gets populated into `mainPackageNotes` and then used to determine whether to show the message origins got lost along the way.

That needs a pain as remember `--hint:MsgOrigin=on` or whatever it is is difficult to remember and really should be a CLI option.